source/adapters/hip/enqueue.cpp

```diff
@@ -519,15 +601,28 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueMemBufferReadRect(
             UR_RESULT_ERROR_INVALID_SIZE);
 
   ur_result_t Result = UR_RESULT_SUCCESS;
-  void *DevPtr = std::get<BufferMem>(hBuffer->Mem).getVoid();
+  ur_lock MemoryMigrationLock(hBuffer->MemoryMigrationMutex);
+  // Note that this entry point may be called on a specific queue that may not
+  // be the last queue to write to the MemBuffer
+  auto DeviceToCopyFrom = hBuffer->LastEventWritingToMemObj == nullptr
+                              ? hQueue->getDevice()
+                              : hBuffer->LastEventWritingToMemObj->getDevice();
+  void *DevPtr = std::get<BufferMem>(hBuffer->Mem).getVoid(DeviceToCopyFrom);
   std::unique_ptr<ur_event_handle_t_> RetImplEvent{nullptr};
 
   try {
-    ScopedContext Active(hQueue->getDevice());
+    ScopedContext Active(DeviceToCopyFrom);
     hipStream_t HIPStream = hQueue->getNextTransferStream();
 
-    Result = enqueueEventsWait(hQueue, HIPStream, numEventsInWaitList,
-                               phEventWaitList);
+    UR_CHECK_ERROR(enqueueEventsWait(hQueue, HIPStream, numEventsInWaitList,
+                                     phEventWaitList));
+    if (hBuffer->LastEventWritingToMemObj != nullptr &&
+        hQueue->getDevice() != DeviceToCopyFrom) {
```

Comment:
This isn't necessarily true. We still need to check both conditions

Reply:
Why? `DeviceToCopyFrom` is effectively:

```cpp
if (LastEvent != null)
  DeviceToCopyFrom = LastEvent->getDevice()
else
  DeviceToCopyFrom = Queue->getDevice()
```

It can only ever be equal to queue's device if last event is not set.
If you want to ensure that last event's device is different to the one from the queue, I think it would make more sense to rewrite the conditional to something like:
```cpp
if (LastEvent && (LastEvent->getDevice() == Queue->getDevice())
```
to encapsulate that logic.

Reply:
Ah sorry I misread your first comment. Yes this is what the check is atm if you are OK with that

Reply:
As in the intention is to make sure that last event's device is different to the queue's device? If so I still think it would read better as:
```cpp
if (LastEvent && (LastEvent->getDevice() != Queue->getDevice())
```

Reply:
Is this not the same as this https://github.com/oneapi-src/unified-runtime/pull/999/files#diff-a70aa00079bd927ead22a6128aa67fba1c6016e919cf7981d722c60d3b111b91R157 ? Do you think I should make a local var for `LastEvent`?

Reply:
Yeah, that's it, I like how you re-wrote the conditional to revolve around the `LastEventWritingToMemObj` in here: https://github.com/oneapi-src/unified-runtime/pull/999/files#diff-a70aa00079bd927ead22a6128aa67fba1c6016e919cf7981d722c60d3b111b91R150
I think this snippet got get the same treatment, it would make it much easier to follow.

Reply:
> Do you think I should make a local var for `LastEvent`?

Probably not needed, modern CPUs are good at chasing pointers.